Reconcile Moran processes

drvinceknight opened the pull request:
Use spatial Moran Processes throughout. 

Also add ability to pass a prob_end to the Matches.

In terms of performance this actually seems faster:

```
%%timeit
# Branch implementation of Moran Process
for players in [[axl.Cooperator()] * 3 + [axl.Defector()] * 3,
                [axl.Random(), axl.Grudger(), axl.TitFor2Tats()],
                [s() for s in axl.demo_strategies]]:

    for seed in range(100):
        axl.seed(seed)
        mp = axl.MoranProcess(players)
        mp.play()
        winner = mp.winning_strategy_name
```

output: `1 loop, best of 3: 8.7 s per loop`

whereas current `master`:

```
%%timeit
# Master branch implementation of Moran Process
for players in [[axl.Cooperator()] * 3 + [axl.Defector()] * 3,
                [axl.Random(), axl.Grudger(), axl.TitFor2Tats()],
                [s() for s in axl.demo_strategies]]:

    for seed in range(100):
        axl.seed(seed)
        mp = axl.MoranProcess(players)
        mp.play()
        winner = mp.winning_strategy_name
```
Output: `1 loop, best of 3: 9.92 s per loop`

This complements #1042.

marcharper commented on docs/tutorials/further_topics/moran_processes_on_graphs.rst:
```diff
@@ -4,7 +4,7 @@ Moran Process on Graphs
 
 The library also provides a graph-based Moran process [Shakarian2013]_ with
-:code:`MoranProcessGraph`.  To use this class you must supply at least one
+:code:`MoranProcess`.  To use this class you must supply at least one
```
class --> feature since the class can be used without a graph.

marcharper commented on axelrod/moran.py:
```diff
@@ -31,9 +32,10 @@ def fitness_proportionate_selection(scores):
+    def __init__(self, players, turns=DEFAULT_TURNS, prob_end=None, noise=0,
                  deterministic_cache=None, mutation_rate=0., mode='bd',
-                 match_class=Match):
+                 match_class=Match, interaction_graph=None,
```
Do we need match_class any longer?

drvinceknight replied:
I don't think so, I'll take it out.

marcharper commented on axelrod/moran.py:
```diff
@@ -101,6 +119,23 @@ def __init__(self, players, turns=DEFAULT_TURNS, noise=0,
+            interaction_graph = complete_graph(len(players), loops=False)
+            if reproduction_graph is None:
+                reproduction_graph = complete_graph(len(players))
+        if reproduction_graph is None:
```
Is this the right default? I think we should consider using the same default in all cases.

drvinceknight replied:
This is the same default as it was implemented in `MoranProcessGraph` but I'll change it. I've got a further idea for how I'd like to refactor this as well.

marcharper commented on axelrod/moran.py:
```diff
@@ -124,15 +159,21 @@ def mutate(self, index):
         return new_player
 
     def death(self, index=None):
-        """Selects the player to be removed. Note that the in the birth-death
```
This docstring is still relevant, no?

drvinceknight replied:
Yup, not sure why I got rid of it.

marcharper commented:
One concern is that without the non-graph implementation we don't have a canonical implementation to compare to. We should consider adding some tests that really demonstrate that it is equivalent (e.g. explicitly tests the interaction pairs for various population sizes).

drvinceknight commented:
> We should consider adding some tests that really demonstrate that it is equivalent (e.g. explicitly tests the interaction pairs for various population sizes).

Sure. I'm going to add to the tests a bit anyway, I'll try to incorporate your suggestion but I might need more guidance. :+1:

drvinceknight commented:
I've just pushed a bunch of changes. Amongst other things I've added numpy docstrings and type hints. I've also gone through and added more unit type tests. I took a look at adding specific interaction tests but I'm not sure it's necessary:

- Each method is now tested (it wasn't before);
- There are some tests that **do** test the expected overall outcomes.

I'm not at all averse to adding more tests if you're happy to spell them out for me (or perhaps they can be left as a further PR). :+1:

marcharper commented on axelrod/moran.py:
```diff
@@ -389,14 +395,15 @@ class ApproximateMoranProcess(MoranProcess):
     from a dictionary of player tuples to distribution of match outcomes
     """
-    def __init__(self, players, cached_outcomes, mutation_rate=0.):
+    def __init__(self, players: List[Player], cached_outcomes: dict,
```
We need all the parameters to pass through?

drvinceknight replied:
Are you asking about the `turns` `prob_end` etc? In which case: no we don't. They are not used as all match results sampled from the `cached_outcomes`.

If you're asking about the `graphs`. I didn't want to tackle that here. It would be an easy thing to add.

drvinceknight replied:
> It would be an easy thing to add.

But let's do that at a later date.

marcharper commented on axelrod/tests/unit/test_moran.py:
```diff
@@ -151,30 +261,15 @@ def test_three_players_with_mutation(self):
         self.assertEqual(populations, mp.populations)
         self.assertEqual(mp.winning_strategy_name, str(axelrod.Defector()))
 
-    def test_standard_fixation(self):
```
Why remove this test?

drvinceknight replied:
Because it was for a mock match class (it was a test for the `match_class` attribute).

marcharper replied:
Actually it tested the constant fitness case (or else, that was the original intent).

drvinceknight replied:
Most of the tests in here were testing multiple things (which I've tried to address). This test also tested that you could pass a different match class and has been removed because I've removed the `match_class` attribute (as you suggested earlier).

drvinceknight replied:
If you wanted to test constant fitness we can do that by passing a cache where all strategies have the same score.

drvinceknight replied:
Actually, I don't understand what that test was doing at all with the Mock class (apart from testing that it worked). I'm fixed the py3.6 problems and about to push with a version of this test (just using a normal match).

marcharper replied:
It was returning a constant score (fitness) in a 2:1 ratio to see if the outcome was in the ballpark of what would be expected for an r=2 Moran process. This appears to be what the new test essentially does.

drvinceknight replied:
Cool :) Let me know if you think there's anything else that we can add.

marcharper commented on docs/tutorials/further_topics/moran_processes_on_graphs.rst:
```diff
@@ -17,8 +17,9 @@ The nodes can be any hashable object (integers, strings, etc.). For example::
-Graphs are undirected by default. Various intermediates such as the list of
-neighbors are cached for efficiency by the graph object.
+Graphs are undirected by default but you can pass :code:`directed=True` to
+create a directed graph.. Various intermediates such as the list of neighbors
```
..

drvinceknight commented:
Failures are py3.6 vs py3.5... Working on basically ensuring the dicts are ordered as expected.

drvinceknight commented:
- 5c26eee gets tests to pass on both py3.6 and py3.5
- b3960ca adds a constant fitness test case
